pkg/scheduler/scheduler_test.go

```diff
@@ -1049,17 +1052,23 @@ func TestFrameworkHandler_IterateOverWaitingPods(t *testing.T) {
 			// Wait all pods in waitSchedulingPods to be scheduled.
 			wg.Wait()
 
-			// Ensure that all waitingPods in scheduler can be obtained from any profiles.
-			for _, fwk := range scheduler.Profiles {
-				actualPodNamesInWaitingPods := sets.NewString()
-				fwk.IterateOverWaitingPods(func(pod framework.WaitingPod) {
-					actualPodNamesInWaitingPods.Insert(pod.GetPod().Name)
-				})
-				// Validate the name of pods in waitingPods matches expectations.
-				if actualPodNamesInWaitingPods.Len() != len(tc.expectPodNamesInWaitingPods) ||
-					!actualPodNamesInWaitingPods.HasAll(tc.expectPodNamesInWaitingPods...) {
-					t.Fatalf("Unexpected waitingPods in scheduler profile %s, expect: %#v, got: %#v", fwk.ProfileName(), tc.expectPodNamesInWaitingPods, actualPodNamesInWaitingPods.List())
+			// When permit plugin emits the event, pod may not been added to the waitingPods pool yet, so we use pollUntil here.
+			if err := wait.PollUntilContextCancel(ctx, 100*time.Microsecond, true, func(context.Context) (done bool, err error) {
+				// Ensure that all waitingPods in scheduler can be obtained from any profiles.
+				for _, fwk := range scheduler.Profiles {
+					actualPodNamesInWaitingPods := sets.NewString()
+					fwk.IterateOverWaitingPods(func(pod framework.WaitingPod) {
+						actualPodNamesInWaitingPods.Insert(pod.GetPod().Name)
+					})
+					// Validate the name of pods in waitingPods matches expectations.
+					if actualPodNamesInWaitingPods.Len() != len(tc.expectPodNamesInWaitingPods) ||
+						!actualPodNamesInWaitingPods.HasAll(tc.expectPodNamesInWaitingPods...) {
+						return false, fmt.Errorf("Unexpected waitingPods in scheduler profile %s, expect: %#v, got: %#v", fwk.ProfileName(), tc.expectPodNamesInWaitingPods, actualPodNamesInWaitingPods.List())
+					}
 				}
+				return true, nil
+			}); err != nil {
+				t.Fatal("got unexpected result")
```

Comment:
Hrrrm. `wait.Poll` is not a good way to poll in tests (see "Good code that waits for something to happen meets the following criteria" in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#polling-and-timeouts).

If this times out, then all that shows up in the unit test output is "got unexpected result". At least include the error, which should be the one from `fmt.Errorf`.

Reply:
You could switch to k/k/test/utils/ktesting and use https://pkg.go.dev/k8s.io/kubernetes/test/utils/ktesting#Eventually. But that's a bigger change.

Up to you...

Reply:
Actually... does `wait.Poll` really poll here? As far as I can tell, your callback is getting invoked *once* and either returns "true" or an error. In both cases [wait.PollUntilContextCancel](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollUntilContextCancel) returns immediately, without retrying later. So this is the same as before, just with a less informative failure?

Reply:
agreed, this seems racy both before and after this change

Reply:
It's a mistake, fixed.

Reply:
PTAL